smeagol.js

```diff
@@ -0,0 +1,160 @@
+	constructor(selected_browser) {
+		this.browser_api;
+		if (SUPPORTED_BROWSERS.firefox === selected_browser) {
+			this.browser_api = browser;
+		} else if (SUPPORTED_BROWSERS.chrome === selected_browser) {
+			this.browser_api = chrome;
+		} else {
+			throw new Error('Smeagol does not support ', selected_browser);
+		}
+	}
+
+	/**
+	 * Send a message to a content script and get its response.
+	 * Firefox - https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/tabs/sendMessage
+	 * Chrome - https://developer.chrome.com/docs/extensions/reference/api/tabs#method-sendMessage
+	 * @param {Number} tabId - Integer ID of the tab to send a message to.
+	 * @param {*} message - Any serializable object to send to a content script.
+	 * @returns {Promise} A Promise that resolves with the response from the
+	 * content script or rejects with an error.
+	 */
+	sendMessageToTab(tabId, message) {
+		return new Promise((resolve, reject) => {
+			this.browser_api.tabs.sendMessage(tabId, message)
+			.then((res) => {
+				res = new MessageResponse({ payload: res.response });
+				if (res.hasError()) {
+					reject(res.getError());
+				}
+				resolve(res.getResponse());
+			}).catch(err => {
+				console.error(err);
+			});
+		});
+	}
+
+	/**
+	 * Send a message to a component of the extension and get its response.
+	 * Firefox - https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/sendMessage
+	 * Chrome - https://developer.chrome.com/docs/extensions/reference/api/runtime#method-sendMessage
+	 * @param {*} message - Any serializable object to send to a content script.
+	 * @returns {Promise} A Promise that resolves with a response from the recipient or rejects with an error.
+	 */
+	sendMessage(message) {
+		return new Promise((resolve, reject) => {
+			this.browser_api.runtime.sendMessage(message)
+			.then((res) => {
+				res = new MessageResponse({ payload: res.response });
+				if (res.hasError()) {
+					reject(res.getError());
+				}
+				resolve(res.getResponse());
+			}).catch(err => {
+				console.error(err);
+			});
+		});
+	}
+
+	/**
+	 * A function that a script will use to respond to messages it recieves.
+	 * @callback MessageHandler
+	 * @param {*} message - Any serializable object to send from one script to another.
+	 * @returns {Promise} A Promise that resolves with a response from the recipient or rejects with an error.
+	 */
+
+	/**
+	 * Designate a MessageHandler for a script using the appropriate API for
+	 * the browser the extension is installed on.
+	 * See the following browser API documentation
+	 * Firefox - https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#addlistener_syntax
+	 * Chrome - https://developer.chrome.com/docs/extensions/reference/api/runtime#event-onMessage
+	 * @param {MessageHandler} messageHandler - A function to handle mesages for
+	 * the script that calls this method.
+	 */
+	addOnMessageListener(messageHandler) {
```

Comment:
It's probably worth referencing the Chrome bug somewhere here. Smeagol would be a lot simpler if we didn't have to manually support promises in Chrome

Reply:
yes and typo on "mesages"